kv.go

```diff
@@ -69,6 +69,9 @@ type KeyValue interface {
 	Bucket() string
 	// PurgeDeletes will remove all current delete markers.
 	PurgeDeletes(opts ...WatchOpt) error
+	// PurgeDeletesOlderThan will remove delete markers that have been present
+	// for longer than the `olderThan` duration.
+	PurgeDeletesOlderThan(olderThan time.Duration, opts ...WatchOpt) error
```

Comment:
We should shoehorn under PurgeDeletes IMO.

Reply:
Would like to, but weird to have the duration value passed as a WatchOpt. What do you suggest then?

Reply:
Creating new Opt that shadows what you could have used reasonably from the watcher which I think is only context IIRC>

Reply:
Just a FYI If it's too broad with too many options you can end up with nonsensical combinations sent to some commands. @scottf highlighted a number of these issues in the past and we narrowed down what watchopts do to avoid that.

Reply:
@ripienaar To make sure I understand what you are saying: we currently pass watch options, which then could allow the user to pass non-sensical options, for instance ignore deletes in which case the functionality would simply not work.

I would not mind replacing WathOpt here with PurgeOpt or something else, but we are all ok to break this? v1.3.0 was released with the PurgeDeletes(opts ...WatchOpt), that is, this is not only in main.

Reply:
As long as we support context which is the only one that would make sense here I think that is correct option.

Reply:
Yea since we have this one size fits all function we pass options in that we use to implement various aspects of the api surface with too broad allowed options on the user facing function you end up with things that just don’t work - not to mention confusion for users. 

I thought we fixed this but can’t find an example now - or Scotts email

Reply:
Would someone, maybe @wallyqs, want to have a go at trying to rework this PR? I find the use of the func options a bit much and can't see how to avoid the stuttering, for instance PurgeDeletes(nats.PurgeDeletesOlderThan()). Of course, the option name should likely be different, but I hope you get the point: I think we will start having issues picking an option functions that:
- are unique enough so that they do not cause conflict with existing or future option function names
- avoid stuttering with the API call that uses them.

In case above, say we shorten the option function name to OlderThan(), it is not clear that this function is for KV purge options, but also would prevent use of that name for an option for other kind of APIs. Think of something more common, say Subject() that could be an option for purging a stream per subject, but could easily also be needed for some kind of other actions, which then would have to be qualified (so that the name is unique - otherwise won't compile), which leads to stuttering.

Reply:
My overall concern was in making sure we did not delete a tombstone that had not yet been delivered to active watchers.

Thinking out loud, maybe first step is to take the existing function and not delete tombstones that are too young? We could still delete the values but keep the tombstone if it is not at least N minutes old?

Reply:
To make sure I understand, suppose we have these keys layout:
```
Key: "foo" - Values: 1, 2, DEL/PURGE
Key: "bar" - Values: 1, DEL/PURGE
```
If "foo"'s DEL/PURGE is say time t-1hour, and "bar"'s DEL/PURGE is t-1minute, a "PurgeDeletes()" API would result on:
```
Key: "foo" - Values: <nothing>
Key: "bar" - Values: DEL/PURGE
```
Everything would be removed from "foo", for "bar", only values (regardless of how "old" they are), we just keep DEL/PURGE that are not old enough. Correct?

Reply:
Btw, I just realized that the current PurgeDeletes() may be flawed if this layout exists (assuming it is possible)
```
Key: "foo" - Values: 1, 2, DEL/PURGE, 3, 4, DEL/PURGE
```
Would the watcher give us 2 del/purge entries for the same key, and we would try to purge the stream by subject twice?

Also, what would be the stream purge request for case where you want to purge a key of its values, but keep the del/purge record? Is it a combination of subject and sequence? If a sequence is given, since it is stream's sequence and messages on various keys (subjects) can interleave, would that work?

Reply:
Hmm, that might be possible, but I think when we do the PurgeDeletes we get last by subject IIRC so we would just get the last DEL/PURGE.

Reply:
I see, so we create a watcher for ">" and don't include history, so we get the last message for each subject/key. Good.

Now the question regarding the PurgeDeletes that should remove data but possibly not the DEL/PURGE record itself (if too young). Is it a matter of doing a purge per-subject along with sequence of this record? My question does the sequence works per-subject (even though it is the stream sequence) or is it going to remove all messages under this sequence number, possibly deleting valid keys, etc.. I am planning to write a test case with foo and bar keys having interleaved (stream sequence wise) sequences and have one of the keep that is supposed to keep the DEL/PURGE, while other not and see if that would work.

Reply:
Not sure if this is what we are looking for still but [in this commit](https://github.com/wallyqs/nats.go/commit/3e0d109b17a3ba155c4134bdd8fc0a9efa57df12) added an `PurgeOpt` that only configures `OlderThan` and `Context` that is then passed to `Watch` for cancellation, plus to avoid the option name being repetitive the option is called `OlderThan` and of `time.Duration` type so that could be used in other contexts potentially if needed.

https://github.com/wallyqs/nats.go/commit/3e0d109b17a3ba155c4134bdd8fc0a9efa57df12

Reply:
@kozlovic the extended purge options allow purge by subject of course but also allow to specify a "keep" number, so in the case where the tombstone is too young I would suggest a purge of the subject with a keep 1, which will be the tombstone.

Reply:
@wallyqs Yes, this is exactly what we would have to do if adding a new option (my comment, though, was that nats.OlderThan() is not really telling what it is for since the function itself is not a receiver of a given object. So if in the future you need an option that has the same semantic, you would need to find a new name. One way is to prefix with the "context", like PurgeDeletes, but then it makes the code look like it stutters: PurgeDeletes(nats.PurgeDeletesOlderThan).

Anyway, I think that would work, but I am looking at Derek's approach to see if it would work. By the way, the threshold to decide if we keep the tombstone or not during a PurgeDeletes could now become a KeyValueConfig struct's field. I will prototype that, and then we can decide which direction we go.

Reply:
I like the idea of threshold being part of KeyValueConfig with sane default, maybe 30minutes by default?

Reply:
That was going to be my next question: so you want to change the default behavior that from now on, the delete/purge markers less than 30min olds will not be removed, unless the user explicitly lower that value. In other words, this new option, if not set, will default to 30min.

Reply:
I think so yes. They can set it, and we do clean up the history of each key regardless.

Reply:
Just leave the tombstone with { keep : 1 } if too young.

Reply:
I have updated the PR (push forced). Let me know if this close to what you had in mind.